Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -113,141 +114,207 @@ end Ideal
+  Iff.rfl
+
+@[simp]
+theorem matrix_apply_stdBasisMatrix [DecidableEq n] {c : RingCon R} {i j : n} {x y : R} :
+    c.matrix n (Matrix.stdBasisMatrix i j x) (Matrix.stdBasisMatrix i j y) ↔ c x y := by
+  refine ⟨fun h ↦ by simpa using h i j, fun h i' j' ↦ ?_⟩
+  obtain hi | rfl := ne_or_eq i i'
+  · simpa [hi] using c.refl 0
+  obtain hj | rfl := ne_or_eq j j'
+  · simpa [hj] using c.refl _
+  simpa using h
+
+theorem matrix_monotone : Monotone (matrix (R := R) n) :=
+  fun _ _ hc _ _ h _ _ ↦ hc (h _ _)
+
+theorem matrix_injective [Nonempty n] : Function.Injective (matrix (R := R) n) :=
+  fun I J eq ↦ RingCon.ext fun r s ↦ by
+    have := congr_fun (DFunLike.congr_fun eq (Matrix.of fun _ _ ↦ r)) (Matrix.of fun _ _ ↦ s)
+    simpa using this
+
+theorem matrix_strictMono_of_nonempty [Nonempty n] :
+    StrictMono (matrix (R := R) n) :=
+  matrix_monotone n |>.strictMono_of_injective <| matrix_injective _
+
+@[simp]
+theorem matrix_bot : (⊥ : RingCon R).matrix n = ⊥ :=
+  eq_bot_iff.2 fun _ _ h ↦ Matrix.ext h
+
+@[simp]
+theorem matrix_top : (⊤ : RingCon R).matrix n = ⊤ :=
+  eq_top_iff.2 fun _ _ _ _ _ ↦ trivial
+
+open Matrix
+
+variable {n}
+
+/-- The congruence relation induced by `c` on `stdBasisMatrix i j`. -/
+def ofMatrix [DecidableEq n] (c : RingCon (Matrix n n R)) : RingCon R where
+  r x y := ∀ i j, c (stdBasisMatrix i j x) (stdBasisMatrix i j y)
+  iseqv.refl _ := fun _ _ ↦ c.refl _
+  iseqv.symm h := fun _ _ ↦ c.symm <| h _ _
+  iseqv.trans h₁ h₂ := fun _ _ ↦ c.trans (h₁ _ _) (h₂ _ _)
+  add' h₁ h₂ := fun _ _ ↦ by simpa [stdBasisMatrix_add] using c.add (h₁ _ _) (h₂ _ _)
+  mul' h₁ h₂ := fun i j ↦ by simpa using c.mul (h₁ i i) (h₂ i j)
+
+@[simp]
+theorem ofMatrix_rel [DecidableEq n] {c : RingCon (Matrix n n R)} {x y : R} :
+    ofMatrix c x y ↔ ∀ i j, c (stdBasisMatrix i j x) (stdBasisMatrix i j y) :=
+  Iff.rfl
+
+@[simp] theorem ofMatrix_matrix [DecidableEq n] [Nonempty n] (c : RingCon R) :
+    ofMatrix (matrix n c) = c := by
+  ext x y
+  classical
+  constructor
+  · intro h
+    inhabit n
+    simpa using h default default default default
+  · intro h i j
+    rwa [matrix_apply_stdBasisMatrix]
+
+end NonUnitalNonAssocSemiring
+
+section NonAssocSemiring
+variable [NonAssocSemiring R] [Fintype n]
+open Matrix
+
+@[simp]
+theorem matrix_ofMatrix [DecidableEq n] (c : RingCon (Matrix n n R)) :
+    matrix n (ofMatrix c) = c := by
+  ext x y
+  classical
+  constructor
+  · intro h
+    rw [matrix_eq_sum_stdBasisMatrix x, matrix_eq_sum_stdBasisMatrix y]
+    refine c.finset_sum _ fun i _ ↦ c.finset_sum _ fun j _ ↦ h i j i j
+  · intro h i' j' i j
+    simpa using c.mul (c.mul (c.refl <| stdBasisMatrix i i' 1) h) (c.refl <| stdBasisMatrix j' j 1)
+
+/-- A version of `ofMatrix_rel` for a single matrix index, rather than all indices. -/
+@[simp]
+theorem ofMatrix_rel' [DecidableEq n] {c : RingCon (Matrix n n R)} {x y : R} (i j : n) :
+    ofMatrix c x y ↔ c (stdBasisMatrix i j x) (stdBasisMatrix i j y) := by
```

Comment:
I see that `coe_equivMatricesOver_symm_apply` corresponds to another spelling of this relation, which is `∃ X Y, c X Y ∧ X i j = x ∧ Y i j = y`, which explains why you can't reuse the proof there.

Reply:
I added that lemma too, but it also doesn't seem to help.